### PR TITLE
Test for racing case of server sending to client on a canceled Subscription 

### PR DESCRIPTION
**Pull request by dymk**
On the client side, rsocket incorrectly would not call `onComplete` or `onError` after the `Subscription` was canceled (in the synchronous case this was deterministic; it races in a multithreading environment however, we should make that clear. This fix does not address multithreading concerns).

I put this test in a separate file because it takes 10+ seconds to build the current RequestStreamTest.cpp file, and linking time isn't slow for us yet.

clang-format results are in a separate commit because it's impossible to filter out the noise otherwise.

### CODE REVIEW

File: test/RequestResponseTest.cpp [hunk 1]
```diff
@@ -157,6 +157,23 @@ TEST(RequestResponseTest, RequestOnDisconnectedClient) {
   ASSERT(did_call_on_error);
 }
 
+// TODO: test that multiple requests on a requestResponse
+// fail in a well-defined way (right now it'd nullptr deref)
+TEST(RequestResponseTest, MultipleRequestsError) {
```
Comment (dymk) [inline test/RequestResponseTest.cpp]: oops, this should go in a separate commit; ignore this

File: test/RequestStreamTest_concurrency.cpp [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+using namespace yarpl;
+using namespace yarpl::flowable;
+using namespace rsocket;
+using namespace rsocket::tests;
+using namespace rsocket::tests::client_server;
+
+struct LockstepBatons {
+  folly::Baton<> on_second_payload_sent;
+  folly::Baton<> on_cancel_sent;
+  folly::Baton<> on_cancel_recieved_toserver;
```
Comment (alexmalyshev) [inline test/RequestStreamTest_concurrency.cpp]: Nit: `s|recieved|received|g`.  Also the usual nag about camelCase.

File: test/RequestStreamTest_concurrency.cpp [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+#define LOCKSTEP_DEBUG(expr) VLOG(3) << expr
+
+class LockstepAsyncHandler : public rsocket::RSocketResponder {
+  LockstepBatons& batons_;
+  Sequence& subscription_seq_;
+
+ public:
+  LockstepAsyncHandler(LockstepBatons& batons, Sequence& subscription_seq)
+      : batons_(batons), subscription_seq_(subscription_seq){};
```
Comment (alexmalyshev) [inline test/RequestStreamTest_concurrency.cpp]: Nit: Rogue semicolon.

File: test/RSocketTests.h [hunk 4]
```diff
@@ -8,6 +8,11 @@
 
 #include "rsocket/transports/tcp/TcpConnectionFactory.h"
 
+auto const default_baton_timeout = std::chrono::milliseconds(100);
+#define CHECK_WAIT(baton) CHECK(baton.timed_wait(default_baton_timeout))
+// #define CHECK_WAIT(baton) baton.timed_wait(default_baton_timeout)
```
Comment (alexmalyshev) [inline test/RSocketTests.h]: I get that they're commented out so they can be easily swapped in, but I don't think it's necessary and clutters things up.

File: test/RequestStreamTest_concurrency.cpp [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+  Sequence server_seq;
+  Sequence client_seq;
+
+  auto server =
+      makeServer(std::make_shared<LockstepAsyncHandler>(batons, server_seq));
+  folly::ScopedEventBaseThread worker;
+  auto client = makeClient(worker.getEventBase(), *server->listeningPort());
+  auto requester = client->getRequester();
+
+  auto scbr_mock =
```
Comment (alexmalyshev) [inline test/RequestStreamTest_concurrency.cpp]: Nit: Just name it `subscriber`, it's clearer and we can afford the vowels :)

File: rsocket/statemachine/ConsumerBase.cpp [hunk 6]
```diff
@@ -45,11 +47,14 @@ void ConsumerBase::generateRequest(size_t n) {
 }
 
 void ConsumerBase::endStream(StreamCompletionSignal signal) {
+  VLOG(5) << "ConsumerBase::endStream(" << signal << ")";
```
Comment (alexmalyshev) [inline rsocket/statemachine/ConsumerBase.cpp]: :+1:

File: test/RSocketTests.h [hunk 7]
```diff
@@ -8,6 +8,9 @@
 
 #include "rsocket/transports/tcp/TcpConnectionFactory.h"
 
+auto const default_baton_timeout = std::chrono::milliseconds(100);
```
Comment (alexmalyshev) [inline test/RSocketTests.h]: Nit: `constexpr`

File: test/RequestResponseTest.cpp (hunk 1 above)
Comment (alexmalyshev) [reply test/RequestResponseTest.cpp]: So is this being removed?
